pkg/server/registry/apigroups/acorn/jobs/restart.go

```diff
@@ -0,0 +1,118 @@
+	ri, _ := request.RequestInfoFrom(ctx)
+
+	if ri.Namespace == "" {
+		return obj, nil
+	}
+
+	// Find the job to restart by looking it up with the public name. The only reason we don't set the actual name
+	// to be the public name is because the public name includes a "." which is not allowed in k8s names.
+	job := &apiv1.Job{}
+	err := s.client.Get(ctx, kclient.ObjectKey{Namespace: ri.Namespace, Name: obj.GetLabels()[labels.AcornPublicName]}, job)
+	if err != nil {
+		return nil, err
+	}
+
+	key := kclient.ObjectKey{Namespace: job.Status.JobNamespace, Name: job.Status.JobName}
+
+	cronToRestart := &batchv1.CronJob{}
+	if err = s.client.Get(ctx, key, cronToRestart); err == nil {
+		return obj, s.restartCronJob(ctx, cronToRestart)
+	}
+
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	jobToRestart := &batchv1.Job{}
+	err = s.client.Get(ctx, key, jobToRestart)
+	if err != nil {
+		return nil, err
+	}
+
+	// Delete the Job and set the propagation policy to foreground so that the dependent resources (Pods)
+	// created by the Job are also deleted.
+	opts := &client.DeleteOptions{PropagationPolicy: z.Pointer(metav1.DeletePropagationForeground)}
+	return obj, s.client.Delete(ctx, jobToRestart, opts)
+}
+
+func (s *restartStrategy) restartCronJob(ctx context.Context, cron *batchv1.CronJob) error {
+	// Find all active jobs and delete them to make way for new ones.
+	for _, jobRef := range cron.Status.Active {
+		job := &batchv1.Job{}
+		err := s.client.Get(ctx, kclient.ObjectKey{Name: jobRef.Name, Namespace: jobRef.Namespace}, job)
+		if err != nil {
+			return err
+		}
+
+		// Delete the Job and set the propagation policy to foreground so that the dependent resources (Pods)
+		// created by the Job are also deleted.
+		opts := &client.DeleteOptions{PropagationPolicy: z.Pointer(metav1.DeletePropagationForeground)}
+		if err := s.client.Delete(ctx, job, opts); err != nil {
+			return err
+		}
+	}
+
+	// Create a new job from the cron's template to "restart" it. This is necessary because
+	// the cronjob controller will not create a new job until the next scheduled run. By creating
+	// it here we can ensure that the job is restarted immediately.
+	template := cron.Spec.JobTemplate
+	spec := template.Spec
+	spec.TTLSecondsAfterFinished = new(int32) // Want to delete the job immediately after it finishes.
+	newJob := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: cron.Name + "-",
+			Namespace:    cron.Namespace,
+			Labels:       template.Labels,
+			Annotations:  template.Annotations,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(cron, batchv1.SchemeGroupVersion.WithKind("CronJob")),
+			},
+		},
+		Spec: spec,
+	}
+	if err := s.client.Create(ctx, newJob); err != nil {
+		return err
+	}
+
+	// Update the cronjob status to reflect the new job being created for it.
+	err := s.client.Get(ctx, kclient.ObjectKey{Name: cron.Name, Namespace: cron.Namespace}, cron)
+	if err != nil {
+		return err
+	}
```

Comment:
You already have the the CronJob via the variable `cron`. Do you need to get it here again?

Reply:
I was mainly afraid of conflict errors here as I'm editing the Jobs that the CronJob creates above. However, testing this has dissuaded my fears as I do not see any conflicts.

Reply:
Additionally, I think we would want conflicts to be returned. For example, what if the conflict means that another run of the job kicked off on its own?